flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java

```diff
@@ -328,4 +325,67 @@ public void operationComplete(ChannelFuture future) throws Exception {
 			}
 		}
 	}
+
+	/**
+	 * Outbound message to be sent to the client.
+	 */
+	public static abstract class ServerOutboundMessage {
+		protected final InputChannelID receiverId;
+		protected final int backlog;
+		private final boolean moreAvailable;
+
+		ServerOutboundMessage(InputChannelID receiverId, int backlog, boolean moreAvailable) {
+			checkArgument(backlog >= 0, "Number of backlog must be non-negative.");
+			this.receiverId = checkNotNull(receiverId);
+			this.backlog = backlog;
+			this.moreAvailable = moreAvailable;
+		}
+
+		abstract Object build();
+
+		public boolean isMoreAvailable() {
+			return moreAvailable;
+		}
+
+		void recycleBufferIfNeeded() {
+		}
+	}
+
+	static class BufferResponseMessage extends ServerOutboundMessage {
+		private final Buffer buffer;
+		private final int sequenceNumber;
+
+		BufferResponseMessage(
+				Buffer buffer,
+				InputChannelID receiverId,
+				int sequenceNumber,
+				int backlog,
+				boolean moreAvailable) {
+			super(receiverId, backlog, moreAvailable);
+			this.buffer = checkNotNull(buffer);
+			this.sequenceNumber = sequenceNumber;
+		}
+
+		@Override
+		Object build() {
+			return new BufferResponse(buffer, sequenceNumber, receiverId, backlog);
+		}
+
+		@Override
+		void recycleBufferIfNeeded() {
+			buffer.recycleBuffer();
+		}
+	}
+
+	static class AddBacklogMessage extends ServerOutboundMessage {
+
+		AddBacklogMessage(InputChannelID receiverId, int backlog, boolean moreAvailable) {
```

Comment:
remove `moreAvailable` argument from the constructor, because it seems strange for `AddBacklogMessage`, then we can give false in below `super` instead.